### PR TITLE
ADD P4 and SAI toolchain

### PR DESCRIPTION
**Pull request by marian-pritsak**
Signed-off-by: Marian Pritsak <marianp@mellanox.com>

### CODE REVIEW

File: sirius-pipeline/README.md [hunk 1]
```diff
@@ -1,2 +1,43 @@
-# Sirius Pipeline P4 Behavior Models
-**TODO**
\ No newline at end of file
+# Sirius Pipeline
+
+## Build the environment
+```
+make docker base
```
Comment (chrispsommers) [inline sirius-pipeline/README.md:5]: Should this be `make docker-base`?

File: SAI/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  apiname               Name of the new SAI API
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --print-sai-lib PRINT_SAI_LIB
+  --sai-git-url SAI_GIT_URL
+  --ignore-tables IGNORE_TABLES
+                        Coma separated list of tables to ignore
+  --sai-git-branch SAI_GIT_BRANCH
+```
```
Comment (chrispsommers) [inline SAI/README.md:21]: Could you please provide an example with valid arguments? Thanks.

Comment (chrispsommers) [reply SAI/README.md:21]: Got error trying to run with just `-h`, looks like some dependencies are needed. Could you provide the appropriate commands or a script? Thanks. See below:
```
DASH-1/SAI$ ./sai_api_gen.py -h
Traceback (most recent call last):
  File "./sai_api_gen.py", line 6, in <module>
    from git import Repo
ModuleNotFoundError: No module named 'git'
```

File: SAI/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+positional arguments:
+  filepath              Path to P4 program BMV2 JSON file
+  apiname               Name of the new SAI API
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --print-sai-lib PRINT_SAI_LIB
+  --sai-git-url SAI_GIT_URL
+  --ignore-tables IGNORE_TABLES
+                        Coma separated list of tables to ignore
```
Comment (chrispsommers) [inline SAI/README.md:19]: Typo: s/coma/comma

File: SAI/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+positional arguments:
+  filepath              Path to P4 program BMV2 JSON file
+  apiname               Name of the new SAI API
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --print-sai-lib PRINT_SAI_LIB
+  --sai-git-url SAI_GIT_URL
+  --ignore-tables IGNORE_TABLES
```
Comment (chrispsommers) [inline SAI/README.md:18]: Are the tables being ignored from the standard SAI tables (e.g. underlay)? Can you provide a default value?

File: SAI/README.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+P4 SAI API generator
+
+positional arguments:
+  filepath              Path to P4 program BMV2 JSON file
+  apiname               Name of the new SAI API
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --print-sai-lib PRINT_SAI_LIB
```
Comment (chrispsommers) [inline SAI/README.md:16]: What does this do?

File: sirius-pipeline/README.md [hunk 6]
```diff
@@ -1,2 +1,43 @@
...
+make bm
```
Comment (chrispsommers) [inline sirius-pipeline/README.md:6]: This fails for me:
```
sirius-pipeline$ make bm
docker run -it --name simple_switch -v /home/chris/nvidia/DASH-1/sirius-pipeline/bmv2:/bmv2 -v /home/chris/nvidia/DASH-1/sirius-pipeline/behavioral-model:/bm-src --network=host --rm bmv2-base bash -c 'cd /bm-src && ./autogen.sh && ./configure && make'
bash: ./autogen.sh: No such file or directory
Makefile:36: recipe for target 'bm' failed
make: *** [bm] Error 127
```
When I look inside the ./behavioral-model directory which is volume-mounted to `bm-src`, I see it is empty. I also noticed the Dockerfile.base has this code:
```
RUN git clone https://github.com/marian-pritsak/behavioral-model && \
        cd behavioral-model && \
	bash install_deps.sh && \
        cd .. && \
        rm -rf behavioral-model
```
The last line appears to delete you cloned, was this intentional?

Comment (budgrise) [reply sirius-pipeline/README.md:6]: I encountered the same issue -- "make bm" fails.  Maybe some files are missing from the PR?

File: SAI/sai_api_gen.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        f.write(''.join(new_lines))
+
+
+# CLI
+parser = argparse.ArgumentParser(description='P4 SAI API generator')
+parser.add_argument('filepath', type=str, help='Path to P4 program BMV2 JSON file')
+parser.add_argument('apiname', type=str, help='Name of the new SAI API')
+parser.add_argument('--print-sai-lib', type=bool)
+parser.add_argument('--sai-git-url', type=str, default='https://github.com/Opencomputeproject/SAI')
+parser.add_argument('--ignore-tables', type=str, default='', help='Coma separated list of tables to ignore')
```
Comment (chrispsommers) [inline SAI/sai_api_gen.py:217]: Can you provide a default list per above comment?

File: SAI/sai_api_gen.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        table_control, table_name = table[NAME_TAG].split('.', 1)
+        sai_table_data['name'] = table_name.replace('.' , '_')
+
+        if sai_table_data['name'] in ignore_tables:
+            continue
+
+        for key in table[KEY_TAG]:
+            sai_table_data['keys'].append(get_sai_key_data(program, key))
+
+        for action in table['actions']:
```
Comment (chrispsommers) [inline SAI/sai_api_gen.py:149]: Can you extract the P4 default action and render a suitable comment in the generated header file? For example, ACL default actions are `deny` and this should appear in the generated .h. Likewise, if there are other attributes or subtleties which are buried or implied by the P4 code, can we try to make them explicit in the generated .h header? If you cannot add to the current PR, perhaps create an issue for this so we can track it. Thanks.

File: sirius-pipeline/README.md [hunk 9]
```diff
@@ -1,2 +1,41 @@
...
+
+## Build the environment
+```
+make docker
+```
+
+## Build pipeline
+```
+make clean
+bmv2/sirius_pipeline.bmv2/sirius_pipeline.json
```
Comment (chrispsommers) [inline sirius-pipeline/README.md:11]: Shouldn't there be a backslash `\` joining these two lines? They are both make targets. I get an error trying to "execute" the second line.

File: sirius-pipeline/README.md [hunk 10]
```diff
@@ -1,2 +1,41 @@
...
+```
+
+## Run software switch
+```
+make run_switch
```
Comment (chrispsommers) [inline sirius-pipeline/README.md:16]: I get errors:
```
DASH-1/sirius-pipeline$ make run-switch 
docker run -it --name simple_switch -v /home/chris/nvidia/DASH-1/sirius-pipeline/bmv2:/bmv2 -v /home/chris/nvidia/DASH-1/sirius-pipeline/behavioral-model:/bm-src --network=host --rm bmv2-chris simple_switch --log-console --interface 0@veth0 --interface 1@veth2 /bmv2/sirius_pipeline.bmv2/sirius_pipeline.json
Calling target program-options parser
Invalid match type: 'list'
Makefile:21: recipe for target 'run-switch' failed
make: *** [run-switch] Error 1
```

File: SAI/README.md [hunk 11]
```diff
@@ -0,0 +1,21 @@
+```
+usage: sai_api_gen.py [-h] [--print-sai-lib PRINT_SAI_LIB]
```
Comment (chrispsommers) [inline SAI/README.md:2]: I tried it twice in a row and got errors:
```
SAI$ ./sai_api_gen.py ../sirius-pipeline/bmv2/sirius_pipeline.bmv2/sirius_pipeline.json dash
Generating SAI API...
Cloning SAI repository...
#
# Repeat
#
SAI$ ./sai_api_gen.py ../sirius-pipeline/bmv2/sirius_pipeline.bmv2/sirius_pipeline.json dash
Generating SAI API...
Cloning SAI repository...
Traceback (most recent call last):
  File "./sai_api_gen.py", line 235, in <module>
    Repo.clone_from(args.sai_git_url, './SAI', branch=args.sai_git_branch)
  File "/home/chris/.local/lib/python3.6/site-packages/git/repo/base.py", line 1111, in clone_from
    return cls._clone(git, url, to_path, GitCmdObjectDB, progress, multi_options, **kwargs)
  File "/home/chris/.local/lib/python3.6/site-packages/git/repo/base.py", line 1049, in _clone
    finalize_process(proc, stderr=stderr)
  File "/home/chris/.local/lib/python3.6/site-packages/git/util.py", line 370, in finalize_process
    proc.wait(**kwargs)
  File "/home/chris/.local/lib/python3.6/site-packages/git/cmd.py", line 447, in wait
    raise GitCommandError(remove_password_if_present(self.args), status, errstr)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v --branch=master https://github.com/Opencomputeproject/SAI ./SAI
  stderr: 'fatal: destination path './SAI' already exists and is not an empty directory.
'
```
Script should not try to clone repo if `SAI` directory already exists.

File: sirius-pipeline/README.md (hunk 1 above)
Comment (marian-pritsak) [reply sirius-pipeline/README.md]: Removed docker-base stage altogether

File: SAI/README.md (hunk 3 above)
Comment (marian-pritsak) [reply SAI/README.md:19]: Done

File: sirius-pipeline/README.md (hunk 6 above)
Comment (marian-pritsak) [reply sirius-pipeline/README.md]: bm stage is removed

File: sirius-pipeline/README.md (hunk 9 above)
Comment (marian-pritsak) [reply sirius-pipeline/README.md]: Fixed

File: sirius-pipeline/README.md (hunk 10 above)
Comment (marian-pritsak) [reply sirius-pipeline/README.md:16]: This match type is not yet supported by the simulator - I will provide another update to have a version of p4 code that can be compiled into the current BMv2.

File: SAI/README.md (hunk 2 above)
Comment (marian-pritsak) [reply SAI/README.md:21]: Added example

File: SAI/README.md (hunk 4 above)
Comment (marian-pritsak) [reply SAI/README.md:18]: I provided a list of tables in the example. I'd prefer not to include them in the script itself.

File: SAI/README.md (hunk 5 above)
Comment (marian-pritsak) [reply SAI/README.md:16]: It's a debug option

File: SAI/sai_api_gen.py (hunk 7 above)
Comment (marian-pritsak) [reply SAI/sai_api_gen.py]: Default is an empty list. I provided tables to ignore in the example for sirius pipeline.

File: SAI/sai_api_gen.py (hunk 8 above)
Comment (marian-pritsak) [reply SAI/sai_api_gen.py:149]: I believe it is possible, but it is outside of the scope of this commit. Can you open an issue?

File: SAI/README.md (hunk 11 above)
Comment (marian-pritsak) [reply SAI/README.md:2]: Added error handling for the case when directory already exists.

File: SAI/sai_api_gen.py (hunk 8 above)
Comment (chrispsommers) [reply SAI/sai_api_gen.py:153]: Done, see https://github.com/Azure/DASH/issues/47

**Comment by chrispsommers** [comment]
Hi @marian-pritsak it looks like you can resolve a trivial README conflict and we can move this forward.

**Pull request merged**